File: 2020/cpp/day_16.cpp
#include "day_16.h"

#include <iostream>

#include <ctre.hpp>
#include <range/v3/all.hpp>

bool day_16::range_t::contains(int i) const {
    const auto in1 = (i >= low1 && i <= high1);
    const auto in2 = (i >= low2 && i <= high2);

    return in1 || in2;
}

day_16::day_16() : input_strings{readFile("Inputs/day_16.txt")} {
    const auto start = std::chrono::high_resolution_clock::now();
    auto sections = split(input_strings, std::string(""));

    for (const auto& line : sections[0]) {
        auto m = ctre::match<R"(^([\w ]+): (\d+)-(\d+) or (\d+)-(\d+))">(line);
        fields[m.get<1>().to_view()] = {fast_atol(m.get<2>().to_view().data()), fast_atol(m.get<3>().to_view().data()),
                                        fast_atol(m.get<4>().to_view().data()), fast_atol(m.get<5>().to_view().data())};
    }

    for (const auto num : split(sections[1][1], ',')) {
        your_ticket.push_back(fast_atol(num.data()));
    }

    for (int i = 1; i < sections[2].size(); ++i) {
        auto& ticket = nearby_tickets.emplace_back();
        for (const auto num : split(sections[2][i], ',')) {
            ticket.push_back(fast_atol(num.data()));
        }
    }
    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("Parsing input for day sixteen took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

bool day_16::isTicketValid(const std::vector<int>& ticket) const {
    auto rng = ranges::views::filter(ticket, [this](auto field) { return isValueValidAnywhere(field); });
    return std::distance(std::begin(rng), std::end(rng)) == ticket.size();
}

bool day_16::isValueValidAnywhere(int field) const {
    int failed{0};
    for (const auto [field_name, field_range] : fields) {
        if (!field_range.contains(field)) {
            ++failed;
        }
    }
    return failed != fields.size();
}

void day_16::part_one() const {
    const auto start = std::chrono::high_resolution_clock::now();
    auto count{0};
    for (const auto& ticket : nearby_tickets) {
        for (auto field : ticket) {
            if (!isValueValidAnywhere(field)) {
                count += field;
            }
        }
    }
    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day sixteen part one is {}\n", count);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

void day_16::part_two() const {
    const auto start = std::chrono::high_resolution_clock::now();
    const auto valid_tickets =
        ranges::views::filter(nearby_tickets, [this](const std::vector<int>& ticket) { return isTicketValid(ticket); }) | ranges::to_vector;
    std::vector<std::pair<std::vector<int>, std::string_view>> candidates;
    for (const auto [field_name, field_range] : fields) {
        std::vector<int> possible;
        for (int i = 0; i < your_ticket.size(); ++i) {
            bool passes = true;
            for (const auto& ticket : valid_tickets) {
                if (!field_range.contains(ticket[i])) {
                    passes = false;
                    break;
                }
            }
            if (passes) {
                possible.push_back(i);
            }
        }
        candidates.emplace_back(possible, field_name);
    }

    std::ranges::sort(candidates, [](auto a, auto b) { return a.first.size() < b.first.size(); });

    robin_hood::unordered_map<std::string_view, int> field_and_index;
    std::vector<bool> taken(your_ticket.size(), false);
    for (const auto& candidate : candidates) {
        for (int index : candidate.first) {
            if (!taken[index]) {
                field_and_index[candidate.second] = index;
                taken[index] = true;
            }
        }
    }
    int64_t result{1};
    for (const auto [field, index] : field_and_index) {
        if (field.find("departure") != std::string_view::npos) {
            result *= your_ticket[index];
        }
    }

    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day sixteen part two is {}\n", result);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

File: 2020/rust/src/day16.rs
use std::{
    collections::HashMap,
    time::{Duration, Instant},
};

use regex::Regex;

type Ticket = Vec<usize>;

#[derive(Debug)]
struct Range {
    low_1: usize,
    high_1: usize,
    low_2: usize,
    high_2: usize,
}

#[derive(Debug)]
struct Parsed<'a> {
    fields: HashMap<&'a str, Range>,
    your_ticket: Vec<usize>,
    nearby_tickets: Vec<Ticket>,
}

impl Range {
    fn contains(&self, i: usize) -> bool {
        (i >= self.low_1 && i <= self.high_1) || (i >= self.low_2 && i <= self.high_2)
    }
}

fn parse_input(input: &str) -> (Parsed, Duration) {
    let start = Instant::now();

    let group_0_regex = Regex::new(r"^([\w ]+): (\d+)-(\d+) or (\d+)-(\d+)").unwrap();
    let lines = input.lines().collect::<Vec<_>>();
    let groups = lines.split(|x| x.is_empty()).collect::<Vec<_>>();

    let fields = groups[0]
        .iter()
        .map(|&line| {
            let capts = group_0_regex.captures(line).unwrap();
            let name = capts.get(1).unwrap().as_str();
            let rng = Range {
                low_1: capts[2].parse().unwrap(),
                high_1: capts[3].parse().unwrap(),
                low_2: capts[4].parse().unwrap(),
                high_2: capts[5].parse().unwrap(),
            };
            (name, rng)
        })
        .collect();

    let your_ticket = groups[1][1]
        .split(',')
        .map(|x| x.parse().unwrap())
        .collect();

    let nearby_tickets = groups[2]
        .iter()
        .skip(1)
        .map(|ticket| ticket.split(',').map(|num| num.parse().unwrap()).collect())
        .collect();
    //let mut nearby_tickets = Vec::new();

    //for i in 1..groups[2].len() {
    //nearby_tickets.push(
    //groups[2][i]
    //.split(',')
    //.map(|num| num.parse().unwrap())
    //.collect(),
    //);
    //}

    let out = Parsed {
        fields,
        your_ticket,
        nearby_tickets,
    };

    let end = Instant::now();
    (out, end - start)
}

fn is_valid_possible_value(input: &Parsed, value: usize) -> bool {
    !input.fields.iter().all(|(_, range)| !range.contains(value))
}

fn is_valid_ticket(input: &Parsed, ticket: &[usize]) -> bool {
    ticket.iter().all(|&x| is_valid_possible_value(input, x))
}

fn solve_part1(input: &Parsed) -> (usize, Duration) {
    let start = Instant::now();
    let mut count = 0;
    for ticket in input.nearby_tickets.iter() {
        for &x in ticket {
            if !is_valid_possible_value(input, x) {
                count += x;
            }
        }
    }
    let end = Instant::now();
    (count, end - start)
}

fn solve_part2(input: &Parsed) -> (usize, Duration) {
    let start = Instant::now();

    let valid_tickets = input
        .nearby_tickets
        .iter()
        .filter(|t| is_valid_ticket(input, t))
        .collect::<Vec<_>>();

    let mut candidates: Vec<(Vec<usize>, &str)> = input
        .fields
        .iter()
        .map(|(&name, range)| {
            (
                (0..input.your_ticket.len())
                    .filter(|i| {
                        valid_tickets
                            .iter()
                            .all(|ticket| range.contains(ticket[*i]))
                    })
                    .collect(),
                name,
            )
        })
        .collect();

    candidates.sort_unstable_by(|(a, _), (b, _)| a.len().cmp(&b.len()));

    let mut field_and_index = HashMap::new();
    let mut taken = vec![false; input.your_ticket.len()];

    for (possible, name) in candidates {
        for index in possible {
            if !taken[index] {
                field_and_index.insert(name, index);
                taken[index] = true;
                break;
            }
        }
    }

    let out = field_and_index
        .iter()
        .filter(|(name, _)| name.starts_with("departure"))
        .map(|(_, idx)| input.your_ticket[*idx])
        .product();

    let end = Instant::now();
    (out, end - start)
}

pub fn run(input: String) {
    let (parsed_input, time) = parse_input(&input);
    println!("Day 16, parsing input took {}ns", time.as_nanos());
    let (p1, time) = solve_part1(&parsed_input);
    println!("Day 16, part 1: {}. Took {}ns", p1, time.as_nanos());
    let (p2, time) = solve_part2(&parsed_input);
    println!("Day 16, part 2: {}. Took {}ns", p2, time.as_nanos());
}
